text/0000-simd-infrastructure.md

```diff
@@ -0,0 +1,404 @@
+                              i0: u32, i1: u32, i2: u32, i3: u32,
+                              i4: u32, i5: u32, i6: u32, i7: u32) -> Simd8<Elem>;
+    fn simd_shuffle16<T, Elem>(v: T, w: T,
+                               i0: u32, i1: u32, i2: u32, i3: u32,
+                               i4: u32, i5: u32, i6: u32, i7: u32
+                               i8: u32, i9: u32, i10: u32, i11: u32,
+                               i12: u32, i13: u32, i14: u32, i15: u32) -> Simd16<Elem>;
+}
+```
+
+The raw definitions are only checked for validity at monomorphisation
+time, ensure that `T` is a SIMD vector, `Elem` is the element type of
+`T` etc. Libraries can use traits to ensure that these will be
+enforced by the type checker too.
+
+This approach has some downsides: `simd_shuffle32` (e.g. `Simd32<u8>`
+on AVX, and `Simd32<u16>` on AVX-512) and especially `simd_shuffle64`
+(e.g. `Simd64<u8>` on AVX-512) are unwieldy. These have similar type
+"safety"/code-generation errors to the vectors themselves.
+
+These operations are semantically:
+
+```rust
+// vector of double length
+let z = concat(v, w);
+
+return [z[i0], z[i1], z[i2], ...]
+```
+
+The indices `iN` have to be compile time constants. Out of bounds
+indices yield unspecified results.
+
+Similarly, intrinsics for inserting/extracting elements into/out of
+vectors are provided, to allow modelling the SIMD vectors as actual
+CPU registers as much as possible:
+
+```rust
+extern "rust-intrinsic" {
+    fn simd_insert<T, Elem>(v: T, i0: u32, elem: Elem) -> T;
+    fn simd_extract<T, Elem>(v: T, i0: u32) -> Elem;
+}
+```
+
+The `i0` indices do not have to be constant. These are equivalent to
+`v[i0] = elem` and `v[i0]` respectively. They are type checked
+similarly to the shuffles.
+
+### Comparisons
+
+Comparisons are implemented via intrinsics, because the current
+comparison operator infrastructure doesn't easily lend itself to
+return vectors, as required.
+
+The raw signatures would look like:
+
+```rust
+extern "rust-intrinsic" {
+    fn simd_eq<T, U>(v: T, w: T) -> U;
+    fn simd_ne<T, U>(v: T, w: T) -> U;
+    fn simd_lt<T, U>(v: T, w: T) -> U;
+    fn simd_le<T, U>(v: T, w: T) -> U;
+    fn simd_gt<T, U>(v: T, w: T) -> U;
+    fn simd_ge<T, U>(v: T, w: T) -> U;
+}
+```
+
+These are type checked during code-generation similarly to the
+shuffles. Ensuring that `T` and `U` has the same length, and that `U`
+is appropriately "boolean"-y. Libraries can use traits to ensure that
+these will be enforced by the type checker too.
+
+### Built-in functionality
+
+Any type marked `repr(simd)` automatically has the `+`, `-` and `*`
+operators work. The `/` operator works for floating point, and the
+`<<` and `>>` ones work for integers.
+
+SIMD vectors can be converted with `as`. As with intrinsics, this is
+"duck-typed" it is possible to cast a vector type `V` to a type `W` if
+their lengths match and their elements are castable (i.e. are
+primitives), there's no enforcement of nominal types.
```

Comment:
What happens if an f32->u32 conversion is out of range?  (This hasn't been resolved for scalars either, but we might pick a different solution for SIMD vectors?)